mithril-common/src/digesters/cardano_immutable_digester.rs

```diff
@@ -49,99 +84,125 @@ impl ImmutableDigester for CardanoImmutableDigester {
-
-                if let Some(cache_provider) = self.cache_provider.as_ref() {
-                    if let Err(error) = cache_provider.store(new_cache_entries).await {
-                        warn!(
-                            self.logger, "Error while storing new immutable files digests to cache";
-                            "error" => ?error
-                        );
-                    }
-                }
+        }
+
+        Ok(digest)
+    }
 
-                Ok(digest)
+    async fn compute_merkle_tree(
+        &self,
+        dirpath: &Path,
+        beacon: &CardanoDbBeacon,
+    ) -> Result<MKTree<MKTreeStoreInMemory>, ImmutableDigesterError> {
+        let immutables_to_process =
+            list_immutable_files_to_process(dirpath, beacon.immutable_file_number)?;
+        info!(self.logger, ">> compute_merkle_tree"; "beacon" => #?beacon, "nb_of_immutables" => immutables_to_process.len());
+        let computed_immutables_digests = self.process_immutables(immutables_to_process).await?;
+
+        let leaves: Vec<_> = computed_immutables_digests
+            .digests
+            .into_iter()
+            .map(|digest| MKTreeNode::new(digest.into()))
+            .collect();
+
+        let mktree =
+            MKTree::new(&leaves).map_err(ImmutableDigesterError::MerkleTreeComputationError)?;
+
+        debug!(
+            self.logger,
+            "Successfully computed Merkle tree for Cardano database"; "beacon" => #?beacon);
+
+        if let Some(cache_provider) = self.cache_provider.as_ref() {
+            if let Err(error) = cache_provider
+                .store(computed_immutables_digests.new_cached_entries)
+                .await
+            {
+                warn!(
+                    self.logger, "Error while storing new immutable files digests to cache";
+                    "error" => ?error
+                );
             }
         }
+
+        Ok(mktree)
+    }
+}
+
+fn list_immutable_files_to_process(
+    dirpath: &Path,
+    up_to_file_number: ImmutableFileNumber,
+) -> Result<Vec<ImmutableFile>, ImmutableDigesterError> {
+    let immutables: Vec<ImmutableFile> = ImmutableFile::list_completed_in_dir(dirpath)?
+        .into_iter()
+        .filter(|f| f.number <= up_to_file_number)
+        .collect();
+
+    match immutables.last() {
+        None => Err(ImmutableDigesterError::NotEnoughImmutable {
+            expected_number: up_to_file_number,
+            found_number: None,
+            db_dir: dirpath.to_owned(),
+        }),
+        Some(last_immutable_file) if last_immutable_file.number < up_to_file_number => {
+            Err(ImmutableDigesterError::NotEnoughImmutable {
+                expected_number: up_to_file_number,
+                found_number: Some(last_immutable_file.number),
+                db_dir: dirpath.to_owned(),
+            })
+        }
+        Some(_) => Ok(immutables),
     }
 }
 
-fn compute_hash(
+fn compute_immutables_digests(
```

Comment:
Maybe this function could be unit tested as the `ComputedImmutablesDigests` struct is not order safe.

IMO a better structure for the `ComputedImmutablesDigests` would be:
```rust
struct ComputedImmutablesDigests {
    entries: BTreeMap<ImmutableFile, HexEncodedDigest>,
    new_cached_entries: Vec<ImmutableFileName>,
}
```

Reply:
In this implementation, we iterate through each value in the `Vec` in the order the elements were inserted. The test `can_compute_hash_of_a_hundred_immutable_file_trio` ensures that the computed hash remains identical to what was previously computed.

I agree with your suggestion to modify the `ComputedImmutablesDigests` struct to use a `BTreeMap`, as it enforces a defined order on `ImmutableFile`. I'll apply this change 👍

Reply:
Done ✅

Reply:
@Alenar proposed this refactoring solution which address both this and the comment above: https://github.com/input-output-hk/mithril/pull/2146/commits/7953b8f68184363f34164f753d821435f64dd006